overlord/hookstate/ctlcmd/kmod.go

```diff
@@ -0,0 +1,197 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package ctlcmd
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/osutil/kmod"
+	"github.com/snapcore/snapd/overlord/hookstate"
+	"github.com/snapcore/snapd/overlord/ifacestate"
+)
+
+var (
+	shortKmodHelp = i18n.G("Load or unload kernel modules")
+	longKmodHelp  = i18n.G(`
+The kmod command handles loading and unloading of kernel modules.`)
+
+	kmodLoadModule   = kmod.LoadModule
+	kmodUnloadModule = kmod.UnloadModule
+)
+
+func init() {
+	addCommand("kmod", shortKmodHelp, longKmodHelp, func() command {
+		cmd := &kmodCommand{}
+		cmd.InsertCmd.kmod = cmd
+		cmd.RemoveCmd.kmod = cmd
+		return cmd
+	})
+}
+
+type kmodNopExecuteMixin struct {
+	kmod *kmodCommand
+}
+
+func (m *kmodNopExecuteMixin) Execute([]string) error {
+	// This is needed in order to implement the interface, but it's never
+	// called.
+	return nil
+}
+
+type kmodCommand struct {
+	baseCommand
+	kmodNopExecuteMixin
+	InsertCmd KModInsertCmd `command:"insert" description:"load a kernel module"`
+	RemoveCmd KModRemoveCmd `command:"remove" description:"unload a kernel module"`
+}
+
+type KModInsertCmd struct {
+	kmodNopExecuteMixin
```

Comment:
And this would have to be replaced with:
```suggestion
	kmod *kmodCommand
```